### DISCUSSION TITLE
Request manipulation fails to update headers

### DISCUSSION THREAD

**Discussion by jfrojo** [chimurai/http-proxy-middleware#593] [Q&A]
I'm manipulating the request (using onProxyReq) performing the following steps:

1.  Parse the request.
2. Insert data into the database via API calls.
3. Update the request (inserting the row's id).
4. Update the headers to reflect the new body length => Fail

When I'm trying to update the proxy request's header it fails due:
` Cannot set headers after they are sent to the client`

My current code looks like this:
```typescript
options.onProxyReq = async (proxyReq: http.OutgoingMessage, req: Request, res: Response): Promise<void> => {
    await registerService(req, res);

    this.fixRequestBody(proxyReq, req, res);
}
``` 

Also if I try to just update the body (and forget about the headers), the error changes to: `ERR_STREAM_WRITE_AFTER_END` meaning the connection is already closed.

I'm doing something wrong? Otherwise, what is the point of enabling the request to be manipulated if you won't wait for the manipulation to happen?

Edit: the fixRequestBody function is https://github.com/chimurai/http-proxy-middleware/blob/master/src/handlers/fix-request-body.ts#L8

**Comment by chimurai** [comment]
Related to this thread: https://github.com/chimurai/http-proxy-middleware/issues/318

Probably easiest way is to create a separate middleware and put it before the proxy middleware:

```js
const manipulate = async (req, res, next) => {
    await registerService(req, res);
    next();
}

app.use('/', manipulate, createProxyMiddleware(...));
```

**Reply by jfrojo** [reply]
My first attempt was like so, but it did fail as the proxy failed to work with the body (I'm using express.json()), so before going back to that approach I wanted to give it a try (using just this framework).

I'll take a look at that thread, thanks for the information!
